### DISCUSSION TITLE
Naming and javadoc clarity

### DISCUSSION THREAD

**Discussion by YouHaveTrouble** [GriefPrevention/GriefPrevention#2258] [General]
I was asked by @RoboMWM to share my rant at how bad the naming in this codebase is. Recently I've been working at implementing [unified protection plugin service api](https://github.com/YouHaveTrouble/YardWatch) and I spend horrendous amount of time trying to figure out which ClaimPermission is responsible for interactions with entities and only after talking to Robo I was informed that `ClaimPermission.Inventory` was the answer. That is straight up not acceptable. There's no mention of the actual responsibility of this enum value anywhere. Either it should have a separate enum value, have its name changed to reflect what it actually represents or at the very least have a javadoc explaining what it's for if there can't be any changes in the enum for any reason.

**Comment by QarthO** [comment]
Can 100% agree with you on this, and I'm sure the other maintainers/contributors will also agree. There are a lot of niche things with the current codebase that have been stitched together and then things added on top of the stitches over the years with new major Minecraft updates and new maintainers of GP. The only reason we're hesitant to change things like this immediately is because of how massively adopted GP is and how it would cause a lot of breakages to the API. I believe the main attraction of GP is how simple and how reliable it is.
So we want to make sure that the new system is setup properly for the future and we give time to developers to support it in their addons/plugins. 

If you take a peep at the roadmap in https://github.com/orgs/GriefPrevention/projects/1 you'll notice that 
v18 the trust system will redone/simplifed, and we'll probably ship out deprecation warnings in the API, 
and in v19 PvE and Api breakage is happening when we'll fully be supporting the changes.

That all being said, there is def a lack of documentation on these things. GP is fully open sourced including the docs, and I know some PRs for the docs would be loved.

Instead of dogging on the how its currently setup, How would you like to see it be implemented?

**Reply by YouHaveTrouble** [reply]
Firstly, update the javadocs on the enum. This won't require major changes in the codebase but will add clarity, even if it will require reading longer javadoc. Will prevent (in my case) literally hours wasted of searching and guessing which enum value is responsible for what.

You mentioned moving away from enums in your other message. I'm not sure if that's a good idea. Moving away from them in favor of registries makes sense because biomes are already a registry and multiple sources can add any amount of them. This isn't really the case with permissions. Sure, you can have more permissions, but there always be a finite amount of them, not sure if other plugins have a reasonable cause for adding a permission, there are finite actions that can be taken and basically everything will fall under those finite actions.

**Reply by QarthO** [reply]
> Firstly, update the javadocs on the enum. This won't require major changes in the codebase but will add clarity, even if it will require reading longer javadoc. Will prevent (in my case) literally hours wasted of searching and guessing which enum value is responsible for what.
> 
> 
> 
> You mentioned moving away from enums in your other message. I'm not sure if that's a good idea. Moving away from them in favor of registries makes sense because biomes are already a registry and multiple sources can add any amount of them. This isn't really the case with permissions. Sure, you can have more permissions, but there always be a finite amount of them, not sure if other plugins have a reasonable cause for adding a permission, there are finite actions that can be taken and basically everything will fall under those finite actions.

Um no, they are switching them to regular classes with static final objects. And yes there can be an infinite amount of permissions. Every plugin can come up with their own permission. For example 

- ClaimPermission.WORLD_EDIT (allowing world edit usage)
- ClaimPermission.VOICE_CHAT (allowing voice chat usage) 
- ClaimPermission.VEINMINER (allowing veinminer usage) 


These are just a random few I can think of off the top of my head. Think of worldguard flags, it's legit an infinite possibility depending on each plugins usecase.

**Reply by YouHaveTrouble** [reply]
Okay, I misunderstood what you meant then, Yea, that's a registry angle.

**Comment by QarthO** [comment]
Our current implementation of the different trust levels is utilizing enums, which is great. but it is limiting addons from creating custom levels.  I think moving away from enums is a good start, similarly to how the Paper team is deciding to move away from enums with things like Biomes

**Comment by Jikoo** [comment]
As pretty much the only person actually putting javadocs on their changes, this feels kinda personal. Yeah, they aren't perfect, but something's better than literally nothing, which is what existed before. Seriously, check the git blame on pretty much any method with documentation actually in Javadoc format. Rather than attack the work that has gone in, maybe you could suggest improvements. On my end, after reading this, I'm feeling like maybe I should just stop bothering to document things if this is the thanks I'll get.

I don't want to make this reply a personal attack on you either, so please forgive me if it does come across that way:
How did you invest what you deem to be a horrendous amount of time searching for entity handling and not search for the EntityDamage(ByEntity)Event? Either of those searches lead you directly to a class which contains exclusively `ClaimPermission.Build` (for armor stands etc. if you trace up) and `ClaimPermission.Inventory`.
I've done my share of searching through nebulous protection plugin code to support them, and GP was one of the easier undocumented plugins to deal with even before I actually started using it on a server and got more familiar with/invested in the codebase. (granted, it was easier because holy heck why are there like only like 3 total 5000 line listeners in the entire plugin, easy to search, oh boy)
